### DISCUSSION TITLE
Middleware or HOCs (Higher Order Components)

### DISCUSSION THREAD

**Discussion by Heet-Bhalodiya** [supabase/supabase#28449] [Questions]
I seek advice on whether to use `middleware` or `Higher Order Components` (HOCs) in my Next.js + Supabase project. Initially, I used `middleware`, but after deploying on the Vercel server, I observed an excessive number of Edge Middleware Invocations. This occurred because the middleware was being triggered with every request. After conducting some research, I opted to implement HOCs, incorporating all the necessary logic within them. This approach ensures that the logic is only executed when the respective component is rendered. I am specifically using `Supabase` authentication in my project and would like to know if utilizing HOCs instead of middleware is a better strategy. Your suggestions would be greatly appreciated. Thank you.

**Comment by ricky-ultimate** [comment]
That depends on the structure of your project. If your project is one that has components rendered based on user level(role i.e, `ADMIN`, `USER` etc) then HOC is highly recommended. But if it is a general display and route protection, then middleware should work just nicely

**Comment by MobileAppVault** [comment]
> I seek advice on whether to use `middleware` or `Higher Order Components` (HOCs) in my Next.js + Supabase project. Initially, I used `middleware`, but after deploying on the Vercel server, I observed an excessive number of Edge Middleware Invocations. This occurred because the middleware was being triggered with every request. After conducting some research, I opted to implement HOCs, incorporating all the necessary logic within them. This approach ensures that the logic is only executed when the respective component is rendered. I am specifically using `Supabase` authentication in my project and would like to know if utilizing HOCs instead of middleware is a better strategy. Your suggestions would be greatly appreciated. Thank you.

@Heet-Bhalodiya  can u give an example on how you build the HOC? I am facing the same issue for my production app.
